File: archive/t/typescript/maximum-subarray.ts
function printUsage(): void {
    console.log('Usage: Please provide a list of integers in the format: "1, 2, 3, 4, 5"');
}

function maxSubarraySum(arr: number[]): number {
    let maxSoFar = Number.MIN_SAFE_INTEGER;
    let maxEndingHere = 0;

    for (let i = 0; i < arr.length; i++) {
        maxEndingHere += arr[i];

        if (maxSoFar < maxEndingHere) {
            maxSoFar = maxEndingHere;
        }

        if (maxEndingHere < 0) {
            maxEndingHere = 0;
        }
    }

    return maxSoFar;
}

function main(): void {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        printUsage();
        process.exit(1);
    }

    // Check if input is empty
    if (args[0].length === 0) {
        printUsage();
        process.exit(1);
    }

    // Parse input string
    const arr = args[0].split(',').map(s => parseInt(s.trim()));

    // If less than two integers were provided
    if (arr.length === 1) {
        console.log(arr[0]);
        return;
    } else if (arr.length < 1) {
        printUsage();
        process.exit(1);
    }

    // Calculate maximum subarray sum
    const result = maxSubarraySum(arr);

    console.log(result);
}

main();

File: archive/e/elixir/maximum-subarray.ex
defmodule MaximumSubarray do

  @doc """
  Returns the maximum subarray sum of the given list using Kadane's algorithm.  
  """
  @spec maximum_subarray(argv :: list(integer())) :: integer()
  def maximum_subarray([]), do: 0
  def maximum_subarray([first | rest]) do
    {_, max_so_far} = Enum.reduce(rest, {first, first}, fn el, {max_ending_here, max_so_far} ->
      max_ending_here = max(el, el+max_ending_here) 
      {max_ending_here, max(max_so_far, max_ending_here)}
    end)
    max_so_far
  end

  @spec parse_list(str :: String.t()) :: {:ok, list(integer())} | :error
  def parse_list(str) do
    str
    |> String.split(",")
    |> Enum.map(&String.trim/1)
    |> Enum.reduce_while({:ok, []}, fn n, {:ok, acc} ->
      case Integer.parse(n) do
        {n, ""} -> {:cont, {:ok, [n | acc]}}
        _       -> {:halt, :error}
      end
    end)
  end
end

with [arg] <- System.argv(),
     {:ok, [_|_] = list} <- MaximumSubarray.parse_list(arg) do
  MaximumSubarray.maximum_subarray(list) |> IO.puts()
else
  _ -> IO.puts("Usage: Please provide a list of integers in the format: \"1, 2, 3, 4, 5\"")
end

